pkg/streamingpromql/operator/binary_operation.go

```diff
@@ -0,0 +1,651 @@
+		sortInterface = favourRightSideSorter{metadata, series}
+	} else {
+		sortInterface = favourLeftSideSorter{metadata, series}
+	}
+
+	sort.Sort(sortInterface)
+}
+
+type favourRightSideSorter struct {
+	metadata []SeriesMetadata
+	series   []*binaryOperationOutputSeries
+}
+
+type favourLeftSideSorter struct {
+	metadata []SeriesMetadata
+	series   []*binaryOperationOutputSeries
+}
+
+func (g favourRightSideSorter) Len() int {
+	return len(g.metadata)
+}
+
+func (g favourLeftSideSorter) Len() int {
+	return len(g.metadata)
+}
+
+func (g favourRightSideSorter) Less(i, j int) bool {
+	iRight := g.series[i].latestRightSeries()
+	jRight := g.series[j].latestRightSeries()
+	if iRight != jRight {
+		return iRight < jRight
+	}
+
+	return g.series[i].latestLeftSeries() < g.series[j].latestLeftSeries()
+}
+
+func (g favourLeftSideSorter) Less(i, j int) bool {
+	iLeft := g.series[i].latestLeftSeries()
+	jLeft := g.series[j].latestLeftSeries()
+	if iLeft != jLeft {
+		return iLeft < jLeft
+	}
+
+	return g.series[i].latestRightSeries() < g.series[j].latestRightSeries()
+}
+
+func (g favourRightSideSorter) Swap(i, j int) {
+	g.metadata[i], g.metadata[j] = g.metadata[j], g.metadata[i]
+	g.series[i], g.series[j] = g.series[j], g.series[i]
+}
+
+func (g favourLeftSideSorter) Swap(i, j int) {
+	g.metadata[i], g.metadata[j] = g.metadata[j], g.metadata[i]
+	g.series[i], g.series[j] = g.series[j], g.series[i]
+}
+
+// hashFunc returns a function that computes the hash of the output group this series belongs to.
+func (b *BinaryOperation) hashFunc() func(labels.Labels) uint64 {
+	buf := make([]byte, 0, 1024)
+	names := b.VectorMatching.MatchingLabels
+
+	if b.VectorMatching.On {
+		slices.Sort(names)
+
+		return func(l labels.Labels) uint64 {
+			var hash uint64
+			hash, buf = l.HashForLabels(buf, names...)
+			return hash
+		}
+	}
+
+	names = append([]string{labels.MetricName}, names...)
+	slices.Sort(names)
+
+	return func(l labels.Labels) uint64 {
+		var hash uint64
+		hash, buf = l.HashWithoutLabels(buf, names...)
+		return hash
+	}
+}
+
+// labelsFunc returns a function that computes the labels of the output group this series belongs to.
+func (b *BinaryOperation) labelsFunc() func(labels.Labels) labels.Labels {
+	lb := labels.NewBuilder(labels.EmptyLabels())
+
+	if b.VectorMatching.On {
+		return func(l labels.Labels) labels.Labels {
+			lb.Reset(l)
+			lb.Keep(b.VectorMatching.MatchingLabels...)
+			return lb.Labels()
+		}
+	}
+
+	return func(l labels.Labels) labels.Labels {
+		lb.Reset(l)
+		lb.Del(b.VectorMatching.MatchingLabels...)
+		lb.Del(labels.MetricName)
+		return lb.Labels()
+	}
+}
+
+func (b *BinaryOperation) NextSeries(ctx context.Context) (InstantVectorSeriesData, error) {
+	if len(b.remainingSeries) == 0 {
+		return InstantVectorSeriesData{}, EOS
+	}
+
+	thisSeries := b.remainingSeries[0]
+	b.remainingSeries = b.remainingSeries[1:]
+
+	allLeftSeries, err := b.leftBuffer.getSeries(ctx, thisSeries.leftSeriesIndices)
```

Comment:
would we gain anything by using a goroutine to get the left+right series at the same time?

Reply:
Potentially query latency would be better, as we'd be able to compute the two sides of the operation in parallel. However, it would mean we'd need to ensure each goroutine doesn't get too far ahead of the other, as otherwise we'd end up holding a bunch of series in memory that we don't need just yet - which would increase peak memory utilisation. 

I think it'd be interesting to explore in the future, but I think it's not necessary right away. Are you OK if I add this to our list of ideas for future improvements?

Reply:
Sure, sounds good to me :+1: